presto-parser/src/main/java/io/prestosql/sql/tree/Window.java

```diff
@@ -20,33 +20,55 @@
 import java.util.Optional;
 
 import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 public class Window
         extends Node
 {
+    private final Optional<Identifier> existingWindowName;
+    private final boolean specification;
```

Comment:
I think we need to do the following:

* Rename `Window` to `Over`. `Over` would have an `Expression` and a `Window` as children.
* `Window` becomes an interface which could be either a `WindowReference` that contains an `Identifier` or a `WindowSpecification`. `WindowSpecification` would contain `existingWindowName`, `partitionBy`, `orderBy` and `frame`.
* `WindowDefinition` would contain an `Identifier` (the name) and a `WiindowSpecification`.

So,

```
class Over
    Expression expression
    Window window

interface Window

class WindowReference implements Window
    Identifier name

class WindowSpecification implements Window
    Identifier existingWindowName
    List<Expression> partitionBy
    Expression orderBy
    WindowFrame frame

class WindowDefinition
    Identifier name
    WindowSpecification specification
```

Reply:
What is `Expression` in class `Over`? 
Why is the current flat structure no good?

Reply:
> What is Expression in class Over?

Sorry, I was thinking about Window/Over as a node that contains a FunctionCall, but it's the other way around.

> Why is the current flat structure no good?

Why I don't like about it is that it mixes two usages into a single union structure. Specifically, it encodes both a simple reference to a window:

* `f() OVER w`
* `f() OVER (w PARTITION BY x ORDER BY y)`

These are very different syntactic structures. In the first one, it's a window function with a window reference. In the second one, it's a window function with an inline specification.

For instance, in the current structure, it would be possible to create a `WindowDefinition` with `Window` that is not a specification programmatically, and that would be syntactically invalid. It'd be equivalent to: `WINDOW x AS y`

Given the above, the structure would be:

```
interface Window

class WindowReference implements Window
    Identifier name

class WindowSpecification implements Window
    Identifier existingWindowName
    List<Expression> partitionBy
    Expression orderBy
    WindowFrame frame

class WindowDefinition
    Identifier name
    WindowSpecification specification
```

So no need for an `Over` class.